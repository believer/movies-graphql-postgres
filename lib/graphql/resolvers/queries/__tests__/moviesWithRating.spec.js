const moviesWithRating = require('../moviesWithRating')
const { db } = require('../../../../adapters/db')

jest.mock('../../../../adapters/db', () => ({
  db: {
    query: jest.fn().mockReturnValue('dbData'),
  },
}))

describe('queries/moviesWithRating', () => {
  beforeEach(() => {
    db.query.mockClear()
  })

  it('should get all movies with rating', () => {
    moviesWithRating(null, { rating: 10 })

    expect(db.query.mock.calls[0][0]).toMatchSnapshot()
  })

  it('should return database data', async () => {
    const response = await moviesWithRating(null, { rating: 10 })

    expect(response).toEqual('dbData')
  })

  it('should handle errors', async () => {
    db.query.mockImplementation(() => Promise.reject('nope'))

    try {
      await moviesWithRating(null, { rating: 9 })
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
