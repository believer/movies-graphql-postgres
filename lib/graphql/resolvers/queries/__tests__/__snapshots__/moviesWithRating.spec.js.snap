// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries/moviesWithRating should get all movies with rating 1`] = `
"
    SELECT 
      distinct on (m.release_date) m.*, 
      
  img.poster, 
  img.backdrop,
  a.actors, 
  d.directors,
  co.composers, 
  w.writers,
  g.genres, 
  c.countries, 
  l.languages, 
  pc.production_companies,
  pro.producers
 
    FROM 
      (
        SELECT * 
        FROM ratings AS r
        WHERE rating = 10
      ) as p,
      LATERAL (
        SELECT * FROM movies WHERE id = p.movie_id
      ) as m, 
      
  LATERAL (
    SELECT * FROM images WHERE movie_id = m.id
  ) as img,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name)
          FROM 
            movie_genres AS comb 
            INNER JOIN genres AS p ON p.id = comb.genre_id
          WHERE 
            comb.movie_id = m.id
        ) as genres
    ) as g 
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name)
          FROM 
            movie_countries AS comb 
            INNER JOIN countries AS p ON p.id = comb.country_id
          WHERE 
            comb.movie_id = m.id
        ) as countries
    ) as c 
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name)
          FROM 
            movie_languages AS comb 
            INNER JOIN languages AS p ON p.id = comb.language_id
          WHERE 
            comb.movie_id = m.id
        ) as languages
    ) as l 
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name)
          FROM 
            movie_production_companies AS comb 
            INNER JOIN production_companies AS p ON p.id = comb.production_company_id
          WHERE 
            comb.movie_id = m.id
        ) as production_companies
    ) as pc 
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 1
        ) as directors
    ) as d
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 2
        ) as actors
    ) as a
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 3
        ) as writers
    ) as w
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 4
        ) as composers
    ) as co
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 5
        ) as producers
    ) as pro
  

    ORDER BY 
        m.release_date DESC;
  "
`;
